lib/_emerge/depgraph.py

```diff
@@ -3686,22 +3673,90 @@ def _wrapped_add_pkg_dep_string(self, pkg, dep_root, dep_priority,
 					if myarg is None:
 						ignored = True
 						dep.child = None
-						self._dynamic_config._ignored_deps.append(dep)
+						ignored_deps.append(dep)
 
 				if not ignored:
 					if dep_priority.ignored and \
 						not self._dynamic_config._traverse_ignored_deps:
 						if is_virt and dep.child is not None:
 							traversed_virt_pkgs.add(dep.child)
 						dep.child = None
-						self._dynamic_config._ignored_deps.append(dep)
+						ignored_deps.append(dep)
 					else:
-						if not self._add_dep(dep,
-							allow_unsatisfied=allow_unsatisfied):
-							return 0
+						if dep.child is None and not dep.blocker:
+							unsatisfied_deps.append(dep)
+						normal_deps.append(dep)
 						if is_virt and dep.child is not None:
 							traversed_virt_pkgs.add(dep.child)
 
+		return _select_atoms_deps(normal_deps, virt_deps, ignored_deps, unsatisfied_deps)
+
+	def _wrapped_add_pkg_dep_string(self, pkg, dep_root, dep_priority,
+		dep_string, allow_unsatisfied):
+
+		debug = "--debug" in self._frozen_config.myopts
+		strict = pkg.type_name != "installed"
+
+		if debug:
+			writemsg_level("\nParent:    %s\n" % (pkg,),
+				noiselevel=-1, level=logging.DEBUG)
+			dep_repr = portage.dep.paren_enclose(dep_string,
+				unevaluated_atom=True, opconvert=True)
+			writemsg_level("Depstring: %s\n" % (dep_repr,),
+				noiselevel=-1, level=logging.DEBUG)
+			writemsg_level("Priority:  %s\n" % (dep_priority,),
+				noiselevel=-1, level=logging.DEBUG)
+
+		autounmask_states = [False]
+		if self._dynamic_config._autounmask:
+			autounmask_states.append(True)
+
+		choices = []
+		for autounmask in autounmask_states:
+			if autounmask:
+				# Clear the package selection cache so that autounmask
+				# can make new selections.
+				self._dynamic_config._filtered_trees[dep_root]["porttree"].dbapi._clear_cache()
+			try:
+				selected_atoms = self._select_atoms(dep_root,
+					dep_string, myuse=self._pkg_use_enabled(pkg), parent=pkg,
+					strict=strict, priority=dep_priority, autounmask=autounmask)
+			except portage.exception.InvalidDependString:
+				if pkg.installed:
+					self._dynamic_config._masked_installed.add(pkg)
+					return 1
+
+				# should have been masked before it was selected
+				raise
+
+			if debug:
+				writemsg_level("Candidates (autounmask=%s): %s\n" % \
+					(autounmask, [str(x) for x in selected_atoms[pkg]],),
+					noiselevel=-1, level=logging.DEBUG)
+
+			choice = self._select_atoms_deps(pkg, dep_root, dep_priority, selected_atoms)
+			choices.append(choice)
+			if not choice.unsatisfied_deps:
+				break
+		else:
+			# If all choices have unsatisfied deps, fall back to default
+			# autounmask=False behavior.
+			choice = choices[0]
+
+			if autounmask:
+				# An autounmask choice has been rejected, so clear its
+				# package selections from the cache.
+				self._dynamic_config._filtered_trees[dep_root]["porttree"].dbapi._clear_cache()
```

Comment:
In the above case we actually need to revert any autounmask settings that might have been created.

Reply:
Also, when autounmask succeeds, it could create extraneous configuration changes. So, there has to be some way to extract only the changes that are actually needed.